File: Data Structures & Algorithms/string-encode-and-decode/submission-7.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        if len(strs) < 1:
            return ""
        
        res = ""
        for s in strs:
            res += str(len(s))
            res += "#"
            res += s
        return res

    def decode(self, s: str) -> List[str]:
        if len(s) < 1:
            return []
        i = 0
        res = []
        while i < len(s):
            j = i
            while s[j] != '#':
                j += 1
            length = int(s[i:j])
            j += 1
            res.append(s[j:j+length])
            i = j + length
        return res

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> chars;
        int i = 0, k = 0;
        int res = 0;
        while(k < s.size()){
            char cur = s[k];
            while(!chars.empty() && chars.find(cur) != chars.end()){
                chars.erase(s[i]);
                i++;
            }

            chars.insert(cur);
            res = max(res, k - i + 1);
            k++;
        }
        return res;
    }
};
